starfish/ddo/starfish_ddo.py

```diff
@@ -0,0 +1,163 @@
+        :param key_id: Key id, Authentication
+        :param authentication_type: Authentication type, str
+        :return: None
+        """
+        if isinstance(key_id, Authentication):
+            # adding an authentication object
+            authentication = key_id
+        elif isinstance(key_id, PublicKeyBase):
+            public_key = key_id
+            # this is going to be a embedded public key
+            authentication = Authentication(public_key, public_key.get_authentication_type())
+        else:
+            # with key_id as a string, we also need to provide the authentication type
+            if authentication_type is None:
+                raise ValueError
+            authentication = Authentication(key_id, authentication_type)
+
+        logger.debug(f'Adding authentication {authentication}')
+        self._authentications.append(authentication)
+
+    def add_signature(self, public_key_store_type=PUBLIC_KEY_STORE_TYPE_PEM, is_embedded=False):
+        """
+        Add signature.
+        Add a signature with a public key and authentication entry for validating this DDO
+        returns the private key as part of the private/public key pair.
+        :param public_key_store_type: Public key store type, str
+        :param is_embedded: bool
+        :return Private key pem, str
+        """
+
+        key_pair = RSA.generate(KEY_PAIR_MODULUS_BIT, e=65537)
+        public_key_raw = key_pair.publickey()
+        private_key_pem = key_pair.exportKey("PEM")
+
+        # find the current public key count
+        next_index = self._get_public_key_count() + 1
+        key_id = f'{self._did}#keys={next_index}'
+
+        public_key = PublicKeyRSA(key_id, owner=key_id)
+
+        public_key.set_encode_key_value(public_key_raw, public_key_store_type)
+
+        if is_embedded:
+            # also add the authentication key as embedded key with the authentication
+            self.add_authentication(public_key)
+        else:
+            # add the public key to the DDO list of public keys
+            self.add_public_key(None, public_key)
+
+            # add the public key id and type for this key to the authentication
+            self.add_authentication(public_key.get_id(), public_key.get_authentication_type())
+
+        logger.debug('Adding signature to the ddo object.')
+        return private_key_pem
+
+
+    def add_public_key(self, did, public_key):
+        """
+        Add a public key object to the list of public keys.
+        :param public_key: Public key, PublicKeyHex
+        """
+        if did == None:
+            self._public_keys.append(public_key)
+        else:
+            logger.debug(f'Adding public key {public_key} to the did {did}')
+            self._public_keys.append(
+                PublicKeyBase(did, **{"owner": public_key, "type": "EthereumECDSAKey"}))
+
+
+    def add_proof(self, authorisation_index, private_key=None, signature_text=None):
```

Comment:
minor quibble about the word "proof". I believe proof is often used with a suffix, e.g. proof_of_XX. In this case, what is XX? It appears (from the code) that the holder of the private key indicates ownership of the private key. Or is it something else?

Reply:
for proof of a valid DDO